model_analyzer/config/generate/automatic_model_config_generator.py

```diff
@@ -0,0 +1,160 @@
+
+    def _step_state_machine(self):
+        if self._done_walking_max_batch_size():
+            self._reset_max_batch_size()
+            self._step_instance_count()
+        else:
+            self._step_max_batch_size()
+
+    def _step_max_batch_size(self):
+        self._curr_max_batch_size *= 2
+
+        last_max_throughput = self._get_last_results_max_throughput()
+        self._curr_max_batch_size_throughputs.append(last_max_throughput)
+
+    def _step_instance_count(self):
+        self._curr_instance_count += 1
+
+    def _done_walking_max_batch_size(self):
+        return self._max_batch_size_limit_reached() \
+            or not self._last_results_increased_throughput()
+
+    def _done_walking_instance_count(self):
+        return self._curr_instance_count >= self._max_instance_count
+
+    def _max_batch_size_limit_reached(self):
+        return self._curr_max_batch_size * 2 > self._max_model_batch_size
+
+    def _last_results_erroneous(self):
+        for measurement in self._last_results:
+            if measurement is None:
+                return True
+        return False
+
+    def _last_results_increased_throughput(self):
+        max_throughput = self._get_last_results_max_throughput()
+        max_throughput_increased = all(
+            max_throughput is not None and max_throughput > t
+            for t in self._curr_max_batch_size_throughputs)
+
+        return max_throughput_increased
+
+    def _reset_max_batch_size(self):
+        if self._sweep_max_batch_size_disabled:
+            self._curr_max_batch_size = self._max_model_batch_size
+        else:
+            self._curr_max_batch_size = self._min_model_batch_size
+        self._curr_max_batch_size_throughputs = []
+
+    def _get_last_results_max_throughput(self):
+
+        throughputs = [
+            m.get_metric_value('perf_throughput')
+            for m in self._last_results
+            if m is not None
+        ]
+        if not throughputs:
+            return None
+        else:
+            return max(throughputs)
+
+    def _get_next_model_config(self):
+        param_combo = self._get_curr_param_combo()
+        model_config = self._make_direct_mode_model_config(param_combo)
+        return model_config
+
+    def _get_curr_param_combo(self):
+
+        if not self._state_machine_started:
+            return self.DEFAULT_PARAM_COMBO
+
+        config = {
+            'dynamic_batching': {},
+            'instance_group': [{
+                'count': self._curr_instance_count,
+                'kind': self._instance_kind
+            }]
+        }
+
+        if not self._sweep_max_batch_size_disabled:
+            config['max_batch_size'] = self._curr_max_batch_size
+
+        return config
+
+    def _determine_sweep_max_batch_size_disabled(self):
+        max_batch_size_disabled = False
+        config = self._get_base_model_config_dict()
+        if "max_batch_size" not in config or config['max_batch_size'] == 0:
```

Comment:
Make this a one-liner (see my above comment) or at least swap the setting of False to be the line directly above